byzcoin/api.go

```diff
@@ -209,13 +245,13 @@ func (c *Client) GetProofFromLatest(key []byte) (*GetProofResponse, error) {
 // state of the chain up to the block. If you need to pass the Proof onwards to
 // another server, you must use GetProof in order to create a complete standalone
 // proof starting from the genesis block.
-func (c *Client) GetProofFrom(key []byte, from *skipchain.SkipBlock) (*GetProofResponse, error) {
+func (c *Client) GetProofFrom(key []byte, from *skipchain.SkipBlock, decoder onet.Decoder) (*GetProofResponse, error) {
```

Comment:
I'm not a big fan of putting all these checks in the receiving method to decide if the received proof is acceptable or not.

We should better define what `SendProtobuf` actually guarantees - triaging the results from different nodes with the `decoder` method looks very flaky to me - sometimes you might get a good reply, sometimes you might not.

E.g., after the call to `getFinalizedParty` you're not really sure if the error is due to the call being too fast, or if the error is due to the fact that the finalized party is not there.

The culprit, IMHO, lies in the `AddTransactionAndWait`. Because, what happens, and what is taken care of not in an ideal way, IMO, in the `getFinalizedParty` case, is that:

1. Transaction is added to a node
2. the node receives a block containing the transaction and returns to the caller
3. the caller asks another node for a proof
4. other nodes still catch up with the latest block

So you get an error in 3., because you ask the proof of a node that did not store the transaction yet (which is normal in a distributed setting).

I think that the `AddTransaction` api needs some flags:
- `WaitForAllNodes` - which makes sure that the block has been processed by all nodes - probably difficult to implement and also flaky
- `ReturnProof` - returns the proof of the instruction - needs to take care which proof should actually be returned
- `ReturnStateChange` - return the statechanges done by the transaction

Reply:
That would be nice to get the proof from the `AddTransactionAndWait` call but you still scenarios where it wouldn't be enough. Like calling a CLI multiple times where we don't share states between the calls so you would still need something. I first thought about using a request to specifically request a proof of existence but it can exist and still be out-of-date.. so that's the best solution I found.

Reply:
CLI: yes, if one command sends the transaction and a next command gets the proof, you're in trouble. But with your solution you'd need to pass the timestamp from the 1st to the 2nd command, so that the second command can correctly call `GetProofAfter`.

Tests/js-frontend/mobile app: here the problem is mostly the `AddTransaction` / `GetProof` combo. In the latest mobile app, I solve the problem by sending the `AddTransaction` and `GetProof` to the same node. So I'm sure that if the `AddTransaction` is done, the `GetProof` is correct, too.

What happens probably most of the time is that `GetProofAfter` accepts the proof from the node that got the `AddTransaction` in the first place (or another node closer to the leader). So it should not be flakey in fact.

Reply:
In fact, `SendProtobufParallel` returns the `network.ServerIdentity` that replied first. So you could use that to send the `GetProof` to, and you'd be sure that the proof includes the latest `ClientTransaction`.

Reply:
It's Friday so let's be crazy:
1. `AddTransaction` request to Bob's node works, everything looks good so far
2. Bob is having a bad day so decides to take his node down
3. `GetProof` to Bob's node fails as it's offline

One of the reason  of `SendProtobufParallel` is to prevent that kind of situation. But yes, you could optimize by asking Bob's node first as you would expect a correct reply from it.

Reply:
OK, looking at `GetProof`, I see that there is `ID`, which can be any block know to the client. So:
1. `AddTransaction` has to return the block where the transaction is actually stored (optional field to be backwards-compatible)
2. `GetProof` gives this block - nodes that are not yet up-to-date will return an error, nodes that do have this block will return the correct proof

Reply:
This should survive even a Friday-crazyness ;)

Reply:
Hm, if you ask for a proof from block `ID`, it will return a proof starting from `ID` which is not always verifiable.

Reply:
True - so we're down to adding a `ReturnProofFrom` flag to the `AddTransaction` message, so that you can get the proof directly if your transaction has been valid. The value of this flag would be used in the same way as the `ID` in `GetProof`.

A more complicated setup would be to add a `ThisBlockNeedsToBeInTheChain` field in the `GetProof` message, but that looks overly complicated and still misses the point if the node did not catch up yet.

Reply:
I was investigating on the last one. We could get the ID from the transaction, and send it to the GetProof so the server is responsible to send back an error if it doesn't know the block. But it can be problematic if a malicious node doesn't stick to the behavior so the client needs to check that the proof indeed contains the block.

Overall that sounds feasible.

Reply:
I think it's OK to say that `AddTransaction` needs to return the proof in such a way that the block containing the transaction needs to be in `Proof.Latest`. Then the client can look at this block and verify whether his transaction is actually in this latest block. And if the rest of the proof checks out, the client can be sure that his transaction has been included in the chain.